File: Android/Basic/AddAnimation/app/src/main/java/com/li/ramon/addanimation/MainActivity.java
package com.li.ramon.addanimation;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Button;

import com.li.ramon.addanimation.frame.FrameAnimActivity;
import com.li.ramon.addanimation.property.LayoutChangesActivity;
import com.li.ramon.addanimation.property.PropertyActivity;
import com.li.ramon.addanimation.view.ViewAnimActivity;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private Button viewAnimation;       // 视图动画
    private Button frameAnimation;      // 帧动画
    private Button propertyAnimation;   // 属性动画

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        init();
    }

    protected void init() {
        viewAnimation = findViewById(R.id.view_animation);
        frameAnimation = findViewById(R.id.frame_animation);
        propertyAnimation = findViewById(R.id.property_animation);

        viewAnimation.setOnClickListener(this);
        frameAnimation.setOnClickListener(this);
        propertyAnimation.setOnClickListener(this);
    }


    @Override
    public void onClick(View view) {
        Intent intent;
        switch (view.getId()) {
            case R.id.view_animation:   // 视图动画
                intent = new Intent(MainActivity.this, ViewAnimActivity.class);
                startActivity(intent);
                break;
            case R.id.frame_animation:  // 帧动画
                intent = new Intent(MainActivity.this, FrameAnimActivity.class);
                startActivity(intent);
                break;
            case R.id.property_animation:   // 属性动画
                intent = new Intent(MainActivity.this, PropertyActivity.class);
                startActivity(intent);
                break;
            case R.id.layout_change:        // 布局改变动画
                intent = new Intent(MainActivity.this, LayoutChangesActivity.class);
                startActivity(intent);
                break;
        }
    }
}

File: Android/View/RemoteViewDemo/app/src/main/java/top/betterramon/remoteviewdemo/MainActivity.kt
package top.betterramon.remoteviewdemo

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.RemoteViews
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat

class MainActivity : AppCompatActivity() {
    private val TAG = "MainActivity"
    private val CHANNEL_ID = "sample_channel"
    private val NOTIFICATION_ID = 1
    private val REQUEST_NOTIFICATION_PERMISSION = 1001
    private val POST_NOTIFICATIONS = "android.permission.POST_NOTIFICATIONS"
    private var  button: Button? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initView()
        createNotificationChannel()
        if (!hasNotificationPermission()) {
            requestNotificationPermission()
        }
    }

    private fun initView() {
        button = findViewById(R.id.showDefaultNotify)!!
        button?.setOnClickListener {
            if (hasNotificationPermission()) {
                showCustomNotification()
            } else {
                requestNotificationPermission()
            }
        }
    }

    private fun requestNotificationPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (ContextCompat.checkSelfPermission(this, android.Manifest.permission.POST_NOTIFICATIONS)
                != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(
                    this,
                    arrayOf(android.Manifest.permission.POST_NOTIFICATIONS),
                    REQUEST_NOTIFICATION_PERMISSION
                )
            }
        }
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val name = "Sample Notification Channel"
            val descriptionText = "Channel for sample notifications"
            val importance = NotificationManager.IMPORTANCE_DEFAULT
            val channel = NotificationChannel(CHANNEL_ID, name, importance).apply {
                description = descriptionText
            }
            val notificationManager =
                getSystemService(NotificationManager::class.java)
            notificationManager.createNotificationChannel(channel)
        }
    }

    private fun showCustomNotification() {
        val remoteViews = RemoteViews(packageName, R.layout.layout_notification)
        remoteViews.setTextViewText(R.id.msg, "Hello from RemoteViews!")
        remoteViews.setImageViewResource(R.id.icon, R.drawable.ic_collected)
        val intent = Intent(this, DemoActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(
            this,
            0,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
        remoteViews.setOnClickPendingIntent(R.id.msg, pendingIntent)

        try {
            val notification = NotificationCompat.Builder(this, CHANNEL_ID)
                .setSmallIcon(R.drawable.ic_launcher_foreground)
                .setContentTitle("Title")
                .setContentText("This is a fallback text.")
                .setCustomContentView(remoteViews) // 设置 RemoteViews
                .setAutoCancel(true)
                .build()

            val notifyManager = NotificationManagerCompat.from(this)
            notifyManager.notify(NOTIFICATION_ID, notification)
        } catch (e: SecurityException) {
            Log.e(TAG, "尝试发送通知时发生异常：${e.message}")
        }
    }

    private fun hasNotificationPermission(): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            ContextCompat.checkSelfPermission(
                this, POST_NOTIFICATIONS
            ) == PackageManager.PERMISSION_GRANTED
        } else true
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if (requestCode == REQUEST_NOTIFICATION_PERMISSION) {
            if ((grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED)) {
                showCustomNotification()
            } else {
                Toast.makeText(this, "未授予通知权限", Toast.LENGTH_SHORT).show()
            }
        }
    }
}
